portal/src/util/error.ts

```diff
@@ -0,0 +1,175 @@
+import { ApolloError } from "@apollo/client";
+import { GraphQLError } from "graphql";
+import { nonNullable } from "./types";
+import { Values } from "@oursky/react-messageformat";
+
+// union type of different kind of violation
+export type Violation = RequiredViolation;
+
+interface RequiredViolation {
+  kind: "required";
+  location: string;
+  missingField: string[];
+}
+
+// list of violation kind recognized
+const violationKinds = ["required"];
+type ViolationKind = Violation["kind"];
+function isViolationKind(value?: string): value is ViolationKind {
+  return value != null && violationKinds.includes(value);
+}
+
+// expected data shape of error extension from backend
+interface ErrorCauseDetails {
+  actual?: string[];
+  expected?: string[];
+  missing?: string[];
+}
```

Comment:
The shape of extensions is defined in `pkg/util/graphqlutil/api_error.go`. All Authgear errors are encoded as API Error.

API Error is a JSON object roughly with the following shape

```typescript
interface APIError {
  errorName: string;
  reason: string;
  // The shape of info depends on reason.
  info: unknown;
}
```

For now, we only need to deal with the reason `ValidationFailed`.

When the reason is `ValidationFailed`, the shape of info is

```typescript
interface ValidationErrorInfo {
  causes: Cause[];
}

interface Cause {
  location: string;
  kind: string;
  // The shape of details depends on kind.
  details: unknown;
}
```

Can we update the type definitions so that the definitions can be extended to cover more `reason` and validation violation `kind` in the future?